react/javascript/demo-external-api/src/components/SheetsProvider.jsx

```diff
@@ -0,0 +1,270 @@
+   * Performs primitive error handling.
+   *
+   * This is a private method.
+   */
+  const invokeSheetsApi = async (url, init) => {
+    try {
+      setError(false)
+      setExpired(false)
+      const { ok, status, body } = await extensionSDK.fetchProxy(url, init)
+      setError(!ok)
+      if (status === 401) {
+        setExpired(true)
+      }
+      return { ok, body, status }
+    } catch (error) {
+      console.error(error)
+      setError(true)
+      return { ok: false }
+    }
+  }
+
+  /**
+   * Unload the currently loaded spreadsheet.
+   */
+  const unloadSpreadSheet = () => {
+    setRows([])
+    setSpreadsheetId()
+    setRange()
+  }
+
+  /**
+   * Load the rows for a given spreadsheet id and range.
+   */
+  const loadSpreadSheet = async (requestSpreadsheetId, requestRange) => {
+    setSpreadsheetId(requestSpreadsheetId)
+    setRange(requestRange)
+    const { ok, body } = await invokeSheetsApi(
+      `https://sheets.googleapis.com/v4/spreadsheets/${requestSpreadsheetId}/values/${requestRange}?access_token=${token}`
+    )
+    if (ok && body && body.values) {
+      setRows(body.values)
+    }
+  }
+
+  /**
+   * Create a new spreadsheet from the source spread sheet provided.
+   */
+  const copySpreadsheet = async (sourceSpreadsheetId) => {
+    let sheetsData
+    {
+      const { ok, body } = await invokeSheetsApi(
+        `https://sheets.googleapis.com/v4/spreadsheets/${sourceSpreadsheetId}?includeGridData=true&access_token=${token}`
+      )
+      sheetsData = ok ? body : undefined
+    }
+    if (sheetsData) {
+      delete sheetsData.spreadsheetId
+      delete sheetsData.spreadsheetUrl
+      let ok
+      let body
+      {
+        // Save sample sheets data to a new sheet.
+        const response = await invokeSheetsApi(
+          `https://sheets.googleapis.com/v4/spreadsheets?access_token=${token}`,
+          {
+            method: 'POST',
+            body: JSON.stringify(sheetsData),
+          }
+        )
+        ok = response.ok
+        body = response.body
+      }
+      if (ok) {
+        return body.spreadsheetId
+      }
+    }
+  }
+
+  /**
+   * Remove a row from the current spreadsheet.
+   */
+  const removeRowFromSpreadSheet = async (rowIndex) => {
+    // row index is based on rows shown. Need to adjust for header.
+    const sourceIndex = rowIndex + 1
+    const { ok } = await invokeSheetsApi(
+      `https://sheets.googleapis.com/v4/spreadsheets/${spreadsheetId}:batchUpdate?access_token=${token}`,
+      {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify({
+          requests: [
+            {
+              deleteDimension: {
+                range: {
+                  sheetId: 0,
+                  dimension: 'ROWS',
+                  startIndex: sourceIndex,
+                  endIndex: sourceIndex + 1,
+                },
+              },
+            },
+          ],
+        }),
```

Comment:
Could `invokeSheetsApi()` do a little more lifting so that the salient parts of the actions is clearer?

Reply:
Yes it could and it now does. I could improve things more but I dont want to re-invent big G's api